client2/thin.go

```diff
@@ -0,0 +1,308 @@
+// merely blocking until the client daemon receives our request to send a message.
+func (t *ThinClient) SendMessage(surbID *[sConstants.SURBIDLength]byte, payload []byte, destNode *[32]byte, destQueue []byte) error {
+	if surbID == nil {
+		return errors.New("surbID cannot be nil")
+	}
+
+	req := new(Request)
+	req.WithSURB = true
+	req.SURBID = surbID
+	req.IsSendOp = true
+	req.Payload = payload
+	req.DestinationIdHash = destNode
+	req.RecipientQueueID = destQueue
+	req.IsSendOp = true
+
+	blob, err := cbor.Marshal(req)
+	if err != nil {
+		return err
+	}
+	count, _, err := t.unixConn.WriteMsgUnix(blob, nil, t.destUnixAddr)
+	if err != nil {
+		return err
+	}
+	if count != len(blob) {
+		return fmt.Errorf("SendMessage error: wrote %d instead of %d bytes", count, len(blob))
+	}
+
+	return nil
+}
+
+func (t *ThinClient) readNextMessage() (*Response, error) {
+	buff := make([]byte, 65536)
+	msgLen, _, _, _, err := t.unixConn.ReadMsgUnix(buff, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	response := Response{}
+	err = cbor.Unmarshal(buff[:msgLen], &response)
+	if err != nil {
+		return nil, err
+	}
+	return &response, nil
+}
+
+// ResponseChan returns the channel that receives message responses
+// from the client daemon, of type ThinResponse.
+func (t *ThinClient) ResponseChan() chan ThinResponse {
+	return t.receivedCh
+}
+
+// ReceiveMessage blocks until a message is received.
+// Use ResponseChan instead if you want an async way to receive messages.
+func (t *ThinClient) ReceiveMessage() (*[sConstants.SURBIDLength]byte, []byte) {
+	resp := <-t.receivedCh
+	return resp.SURBID, resp.Payload
+}
+
+func (t *ThinClient) ARQSend(id *[MessageIDLength]byte, payload []byte, destNode *[32]byte, destQueue []byte) error {
+	req := new(Request)
+	req.ID = id
+	req.WithSURB = true
+	req.IsARQSendOp = true
+	req.Payload = payload
+	req.DestinationIdHash = destNode
+	req.RecipientQueueID = destQueue
+
+	blob, err := cbor.Marshal(req)
+	if err != nil {
+		return err
+	}
+	count, _, err := t.unixConn.WriteMsgUnix(blob, nil, t.destUnixAddr)
+	if err != nil {
+		return err
+	}
+	if count != len(blob) {
+		return fmt.Errorf("SendMessage error: wrote %d instead of %d bytes", count, len(blob))
+	}
+
+	return nil
+}
+
+// ARQReceiveMessage blocks until a message is received.
+// Use ResponseChan instead if you want an async way to receive messages.
+func (t *ThinClient) ARQReceiveMessage() (*[MessageIDLength]byte, []byte) {
+	resp := <-t.receivedCh
+	return resp.ID, resp.Payload
+}
+
+func (t *ThinClient) BlockingSendReliableMessage(id *[MessageIDLength]byte, payload []byte, destNode *[32]byte, destQueue []byte) (reply []byte, err error) {
+	err = t.ARQSend(id, payload, destNode, destQueue)
+	if err != nil {
+		return nil, err
+	}
+	id2, reply := t.ARQReceiveMessage()
+	if !bytes.Equal(id[:], id2[:]) {
+		return nil, errors.New("received unexpected ARQ reply")
```

Comment:
we will hit this if we allow concurrent BlockingSendReliableMessage.
If we don't want that we should have more locking, and/or not expose `ARQReceiveMessage` and `ARQSend`.

Or do some complicated thing where callers to ARQSend can register callback channels.

Or a simpler thing where we register `request_id` in a map to signal that we want to buffer responses in a `map[ID] -> ThinResponse`, and then BlockingSendReliableMessage can call ARQReceiveMessage until it gets that. But then using the API yourself becomes harder. I'm not sure.

Reply:
there is no request id. there are only message IDs and SURB IDs